Fetch AWS Account ID from metadata

Fodoj opened the pull request:
In addition to meta-data and userdata EC2 also provides "dynamic" data (not sure why it's called dynamic though). One extra useful attribute that it has, and that seems to be not available in other sources, is AWS account id (try running curl http://169.254.169.254/latest/dynamic/instance-identity/document). It is possible to fetch account id from the instance that has IAM role attached (by parsing ARN of the role) but for instances without any IAM role this dynamic data is the only way to figure out to which account it belongs to.

lamont-granquist commented on lib/ohai/mixin/ec2_metadata.rb:
```diff
@@ -19,6 +19,7 @@
 
 require "net/http"
 require "socket"
+require "json"
```
should be ffi-yajl

thommay replied:
Please use FFIYajl for JSON handling; see https://github.com/chef/ohai/blob/master/lib/ohai/mixin/gce_metadata.rb#L71 (also this validates that we've got valid JSON)

lamont-granquist commented on lib/ohai/mixin/ec2_metadata.rb:
```diff
@@ -200,6 +201,13 @@ def fetch_userdata
+        api_version = best_api_version
+        return nil if api_version.nil?
+        response = http_client.get("/#{api_version}/dynamic/instance-identity/document/")
+        response.code == "200" ? JSON.parse(response.body) : nil
```
should be `FFI_Yajl::Parser.parse`

lamont-granquist replied:
and the nil response here similarly should have better error handling if we're gonna worry about it

lamont-granquist commented on lib/ohai/mixin/ec2_metadata.rb:
```diff
@@ -200,6 +201,13 @@ def fetch_userdata
 
+      def fetch_dynamic_data
+        api_version = best_api_version
+        return nil if api_version.nil?
```
this is going to trainwreck later with a "NoMethodError on NilClass" and should really handle this better

lamont-granquist commented on lib/ohai/mixin/ec2_metadata.rb:
```diff
@@ -200,6 +201,13 @@ def fetch_userdata
         response.code == "200" ? response.body : nil
       end
 
+      def fetch_dynamic_data
```
also this could be memoized since it looks like you intend it to get reused `@fetch_dynamic_data ||= begin ... end`

Fodoj replied:
I don't think it should be memoized. It's a dynamic data and some values from it can change any time.

Fodoj replied:
Also, other attributes like user data is not memoized as well in this plugin

Fodoj commented:
@lamont-granquist I've replaced json gem with ffi-yajl. Otherwise implementation is same as fetch_user_data few lines above method I've added. I've also replaced default return value with empty hash, instead of nil, so it doesn't explode inside ec2 plugin. Looking forward to your feedback on this. 

PS

I also rebased and squashed commits into one.

Fodoj commented:
@lamont-granquist any update here?

Fodoj commented:
@thommay done! Thanks for your review
